### ISSUE TITLE
Json cannot load objects instances from file.

### ISSUE THREAD

**Issue by GianptDev** [GianptDev/Godot-ConfigBox#1]
# Problem
Currently exist a problem in `Singletone :: _File` which the function `value_to_json` convert the object into a dictionary but the dictionary cannot be converted into a object from `json_to_value` because there is not code for objects.

The main cause is because when reading the file it will get a dictionary and not an object, this mean that all of the code must be added in `TYPE_DICTIONARY` in the type match, but the dictionary is also a different type and there are two options to fix this problem:
- The object is stored in another way.
- The dictionary store some extra information to specify that he is an object.

The object script path can be ignored because if the option is defined whit the object already instanced, then only the values must change.
Maybe storing the path of the script may be useful to know if the object is using the same script just for a check or something.

Also `value_to_json` must be re-executed to allow to store sub-objects and sub-values.

What can i say? Json is really bad, but i'll find a solution just because of the challenge.

**Comment by GianptDev** [comment]
Currently the json will load from file as a dictionary, but sub-values aren't converted properly and this is just a placeholder solution. Added this into the v1.1 project.

**Comment by GianptDev** [comment]
# Solution
Just figured out how to store objects into json.

### Writing in the file
Godot include a function called `inst2dict` which can convert objects into a dictionary of all values from the object + his script path.

The result will give also non exported values and whit a loop i've removed all of them from the dicitonary, the result will then written inside the file.

### Reading from the file
Now, how about restarting the object again?
Godot also include `dict2inst` but it will instantly make a new instance from the script, and if i replace the new instance whit the old one probably some things will not go right.

So i made that when loading the file and instancing the object it will check if an object that use the same script already exist and then get the list of values from the file and just set them to the already existing object.

If the config is null it will use the data inside the file and instance the object as new.

The voilà! now is possible to store any object class inside json and it will set the values if the class is already instanced or will make a new one.

### Final words

Even if the object can now be saved in json all of his value cannot be properly converted into a proper way, but it may fix using `json_to_value` whit some re-execution.

Current version (v1.0) will crash if you try to save into json because of a error inside `json_to_value` which it will try to convert a value from a wrong type and will not actually load anything, now is fixed but this will be included inside v1.1

So basically, until i release v1.1 keep saving into res/tres

**Comment by GianptDev** [comment]
# Results
After a while and a lot of brain cells totally destroyed by coding i have a functional script that convert an object into json and can load the data whitout overriding the existing object.

# More problems
The entire script is now more than 1000 lines long because of the abnormal amount of checks i need to do to prevent crash or data errors, this mean that the code is **really** slow because the object is converted into a dictionary whit information about the object and convert all sub-values, and this slow down even other features.

I've made the same code also for res/tres (which they run much faster) but then i realized something: _Why should someone save a config as an object if is really slow and is basically dictionary which do the exact same job but better?_

# Conclusion
I will not delete the current code i have, but i will not make this update for v1.1

The code is too bad to be considered an update, and as i said dictionary should already do this job.

I will close this issue and i will not add this to the new update, maybe a next update can reconsiderate this feature?

**Comment by GianptDev** [comment]
# Post Scriptum

After some clean-up in the code i maybe found a new way to store objects whit less work by the machine.

Now there is 2 choices how the object need to be loaded:
- if the object need to be reinstanced as new
- if the object (that already exist in memory) need to be updated whit the loaded values

Maybe i will try to do some tests and see if i can make another issues for v1.2
